### ISSUE TITLE
Deep link into app in Expo client does not close the Chrome Custom Tabs browser

### ISSUE THREAD

**Issue by brentvatne** [expo/expo#4524]
### Environment

Android Expo client. This bug does not occur in Android standalone apps.

### Steps to Reproduce

1. Open a web browser with `WebBrowser.openBrowserAsync(url);`
2. Have that URL redirect back to your app (`Linking.makeUrl()`)
3. Notice that in Expo client for Android you need to manually close the browser by tapping the X button, but in a standalone app it closes as expected.

### Expected Behavior

On Android because we cannot call `WebBrowser.dismiss()` we depend on the browser window closing when we redirect back to the app. 

### Actual Behavior

The browser window only closes on redirect in standalone apps.

### Reproducible Demo

```js
import React from 'react';
import { Button, StyleSheet, Text, View } from 'react-native';
import { AuthSession } from 'expo';

const FB_APP_ID = '2232923493627585';

export default class App extends React.Component {
  state = {
    result: null,
  };

  render() {
    return (
      <View style={styles.container}>
        <Button title="Open FB Auth" onPress={this._handlePressAsync} />
        {this.state.result ? (
          <Text>{JSON.stringify(this.state.result)}</Text>
        ) : null}
      </View>
    );
  }

  _handlePressAsync = async () => {
    try {
      let redirectUrl = AuthSession.getRedirectUrl();
      let result = await AuthSession.startAsync({
        authUrl:
          `https://www.facebook.com/v2.8/dialog/oauth?response_type=token` +
          `&client_id=${FB_APP_ID}` +
          `&redirect_uri=${encodeURIComponent(redirectUrl)}`,
      });
      this.setState({ result });
    } catch(e) {
      alert(e.message);
    }
  };
}

const styles = StyleSheet.create({
  container: {
    flex: 1,
    alignItems: 'center',
    justifyContent: 'center',
  },
});
```

(be sure to set a `scheme` for the standalone app when testing there)

**Comment by tonnenpinguin** [comment]
Is there a workaround for this?

Given that this seems to break our SSO workflow I guess there's not way to finally upgrade to 33 at the moment :(

**Comment by d4rky-pl** [comment]
@tonnenpinguin This only happens for apps that are being run in the Expo application, not for the standalone (released) apps.

Unfortunately because it happened all the time when we were QAing or testing the application, we also had to postpone the upgrade to SDK 33 because it could potentially lead to the bugs being introduced and only caught in the production :(

As far as I'm aware, there's no workaround.

**Comment by jonsamp** [comment]
I also see this issue in our standalone production android app, and it's preventing any android device from logging in with an `AuthSession`.

Here's what happens:

1. Click on a button that calls `AuthSession.startAsync`.
2. This summons up a chrome tab, after i sign in, the chrome tab briefly says "Return to app ->". Tapping on that does nothing. Then the chrome tab turns blank.
3. When clicking on the "X" on the chrome tab, then there's an Error that you can't dismiss the window, and the `AuthSession` promise never returns.

This was working fine in SDK 32, but is now broken in SDK 33. This is causing us to remove logging in on android with third party auth with `AuthSession`.

**Comment by mczernek** [comment]
This should be resolving the issue: [#4882](https://github.com/expo/expo/pull/4882)

**Comment by d4rky-pl** [comment]
@mczernek since the fix was merged, any chance you could bump the package version and release it so we could see if it solved the issue? Thanks! :)

**Comment by d4rky-pl** [comment]
@mczernek hey, are there any plans to backport this fix to SDK33 or should we wait for SDK34 before it starts working correctly? We only have 10 days left to do the upgrade and it's getting a bit stressful :)

**Comment by rossb89** [comment]
Well this is annoying! I thought with the release of SDK34 this would fix the issue.. it's still an issue. You still have to close the chrome window manually and it doesn't do it for you like it used to before this bug :(

**Comment by thargenediad** [comment]
So SDK34 doesn't fix this?  We're trying to meet the 8/1 deadline for 64-bit Android apps & having to close the Chrome window manually is not a good user experience.

**Comment by rossb89** [comment]
Not in my testing i've done, performs exactly the same as v33, which is not working. The promise resolves still in the background, but the chrome custom tab stays open until you physically tap to close it.
I can see the code change in #4882 is in the v34 release though, so clearly that hasn't fixed the issue.
Just testing also opening the page with the app redirect link using 'open in chrome' and the pressing it from there, works straight away, so it's definitely the custom tab stuff that's the issue.

**Comment by LuisEgan** [comment]
Hello! I'm also having this issue and it seems that after interacting with *anything* (even tapping in white space) in the `WebBrowser` the deep link won't work, BUT if you do a `window.location.href = goBackURL` and NO interaction before, it does work, closing the browser! You could put it in a `setTimeout()` even to display some message.

Something like this:

In your page.html that you are looking at in the `WebBrowser`

```
<script>
      document.addEventListener("DOMContentLoaded", function(event) {

        // Take the uri from the params
        var qs = decodeURIComponent(document.location.search);
        if (qs) {
          baseUri = qs.split("?linkingUri=")[1];
        }

        var setto = setTimeout(() => {
          var goBackLink =
            baseUri +
            "param=" +
            encodeURIComponent("value");
          window.location.href = goBackLink;
        }, 2000);

        setto();
      });
    </script>
  </head>

  <body>
      ...
  </body>
```

And in your Expo, in the App.js (most top level component)

```
async componentDidMount() {
    if (isAndroid) {
      try {
        let url = await Linking.getInitialURL();
        this._handleRedirect({ url });
      } catch (error) {
        console.log("error: ", error);
      }
    }
    Linking.addEventListener("url", this._handleRedirect);
  }

  componentWillUnmount() {
    Linking.removeEventListener("url", _handleRedirect);
  }

  _handleRedirect(event) {
    if (!isAndroid) WebBrowser.dismissBrowser();

    let data = Linking.parse(event.url);
    console.log("data: ", data);
  }
```

**Comment by thargenediad** [comment]
Thanks, @LuisEgan.  Unfortunately, I don't think this helps those of us using `AuthSession.startAsync` for OAuth purposes.

**Comment by rossb89** [comment]
Hey everyone, so I did some more reading and found a few other projects around which use these same flags to fix the issue of it not closing.

So I thought this surely must be the correct fix...I still can't get it it working when testing in the expo client app, but I did an expo build:android to generate the .apk file with the project using the v34 upgrade, installed that onto the android emulator and the chrome custom tab closed as I hoped it would first time

So potentially i'm thinking this is just now an issue when testing in the expo client app (when using v34). It would be good to know if anyone else could try this as well, get the apk file generated having upgraded to v34 and see if it fixes their issue as well..

**Comment by thargenediad** [comment]
Is this only an issue in the Expo client?  I'd assumed it would behave the same way when published.

**Comment by mczernek** [comment]
Hey! I've just checked the issue, and the problem is that for some reason fix haven't landed in newest Client version. However, when building standalone application it should work just fine. It is also included in expo-web-browser@6.0.0 for ejected and bare applications.

I am on releasing new version of client with that fix for SDK34.

**Comment by mczernek** [comment]
We cannot, temporarily, publish Expo Client update on Google Play Store. Could you please check  [this version](https://d1ahtucjixef4r.cloudfront.net/Exponent-2.12.2.apk) and let us know if it works now?

**Comment by rossb89** [comment]
Just tested on the android emulator with that expo .apk, works a treat, great job guys!

**Comment by d4rky-pl** [comment]
Hey, it's been three weeks, any news on publishing the latest Expo Client on the Google Play Store?

**Comment by mczernek** [comment]
It will be published after SDK 35 is released.

śr., 28 sie 2019 o 14:12 Michał Matyas <notifications@github.com>
napisał(a):

**Comment by superKalo** [comment]
@mczernek maybe a similar issue still persists on my side. I described it [here](https://forums.expo.io/t/expo-linking-fails-on-development-only-on-android/26012) a while ago.

I use Expo Linking for the oAuth flow of my project. The redirect urls (deep links) which the server triggers (and I listen for) towards my app are the ones described in Expo docs:

- Published app in Expo client: exp://exp.host/@community/with-webbrowser-redirect
- Published app in standalone: myapp://
- Development: exp://wg-qka.community.app.exp.direct:80

They all work on iOS (both standalone app, simulator and during dev).

However, on Android, the Development linking fails. I tried to connect my device via a cable and run my project or scan the QR code… the same. Just when a redirection is triggered by my server, every time it fails with the following error:

<img width="468" alt="Screenshot 2019-10-02 at 10 43 30" src="https://user-images.githubusercontent.com/2548061/66026430-82945d00-e501-11e9-8a01-7673c38de3f2.png">

I'm on:
- Expo SDK: v35
- Expo Client: v2.13.0
- Expo Cli: v3.1.2

**Comment by moay** [comment]
@mczernek This seems to still be an issue. We do have a problem related to this when using Facebook oauth. When we send the user to Facebook and the user has already granted all permissions, there will be an immediate redirection to our backend, which in turn redirects to the deeplink in the app. This works fine in the expo client on iOS but fails on Android. The result of `openAuthSessionAsync` (expo-web-browser) will be an object looking like this:
```
{
    type: 'dismiss'
}
```
This object seems to come out of `_openAuthSessionPolyfillAsync` which handles the call on android. Still waiting for testing results on android standalone but I assume that the this issue still isn't released?

Working on expo client 2.14.0

**Comment by moay** [comment]
@brentvatne Thank you very much for taking me seriously. I'm deeply sorry if I caused any major thinking efforts on your side. I don't actually know if this issue persists, but I know for sure that my problems were not related to this issue but to prechristmatic dumbness. I somehow ended up passing in the redirectUrl into `openAuthSessionAsync` instead of the returnUrl. The behaviour I encountered makes total sense then.

I am still mildly surprised that this did not kill the login process on iOS (the returnUrl seems to simply be ignored there?) but thats not part of this issue.
